drivers/rtc/rtc_bq32002.c

```diff
@@ -0,0 +1,370 @@
+#define BQ32002_CAL_PPB_MAX         (31 * BQ32002_CAL_PPB_PER_LSB_NEG)
+
+/* IRQ frequency property enum values */
+#define BQ32002_IRQ_FREQ_ENUM_1HZ      0
+#define BQ32002_IRQ_FREQ_ENUM_512HZ    1
+#define BQ32002_IRQ_FREQ_ENUM_DISABLED 2
+
+/* RTC time fields supported by BQ32002 */
+#define BQ32002_RTC_TIME_MASK                                                                      \
+	(RTC_ALARM_TIME_MASK_SECOND | RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR |      \
+	 RTC_ALARM_TIME_MASK_MONTH | RTC_ALARM_TIME_MASK_MONTHDAY | RTC_ALARM_TIME_MASK_YEAR |     \
+	 RTC_ALARM_TIME_MASK_WEEKDAY)
+
+struct bq32002_config {
+	struct i2c_dt_spec i2c;
+	uint8_t irq_freq;
+};
+
+struct bq32002_data {
+	struct k_sem lock;
+};
+
+static void bq32002_lock_sem(const struct device *dev)
+{
+	struct bq32002_data *data = dev->data;
+
+	(void)k_sem_take(&data->lock, K_FOREVER);
+}
+
+static void bq32002_unlock_sem(const struct device *dev)
+{
+	struct bq32002_data *data = dev->data;
+
+	k_sem_give(&data->lock);
+}
+
+static int bq32002_set_irq_frequency(const struct device *dev)
+{
+	const struct bq32002_config *config = dev->config;
+	uint8_t sf_regs[3];
+	uint8_t cfg1_val;
+	uint8_t cfg2_val;
+	int err;
+
+	switch (config->irq_freq) {
+	case BQ32002_IRQ_FREQ_ENUM_1HZ:
+		cfg1_val = BQ32002_FREQ_TEST_MASK;
+		cfg2_val = BQ32002_FTF_MASK;
+		break;
+	case BQ32002_IRQ_FREQ_ENUM_512HZ:
+		cfg1_val = BQ32002_FREQ_TEST_MASK;
+		cfg2_val = 0;
+		break;
+	default:
+		cfg1_val = BQ32002_OUT_MASK;
+		cfg2_val = 0;
+		break;
+	}
+
+	err = i2c_reg_update_byte_dt(&config->i2c, BQ32002_CAL_CFG1_REG, BQ32002_FREQ_TEST_MASK,
+				     cfg1_val);
+	if (err) {
+		return err;
+	}
+
+	/* Update FTF value if frequency output enabled */
+	if (cfg1_val & BQ32002_FREQ_TEST_MASK) {
+		sf_regs[0] = BQ32002_SF_KEY_1;
+		sf_regs[1] = BQ32002_SF_KEY_2;
+		sf_regs[2] = cfg2_val;
+		err = i2c_burst_write_dt(&config->i2c, BQ32002_SF_KEY_1_REG, sf_regs,
+					 sizeof(sf_regs));
+		if (err) {
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int bq32002_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	const struct bq32002_config *config = dev->config;
+	int err;
+	uint8_t regs[7];
+
+	if ((timeptr == NULL) || !rtc_utils_validate_rtc_time(timeptr, BQ32002_RTC_TIME_MASK) ||
+	    (timeptr->tm_year < BQ32002_YEAR_OFFSET)) {
+		return -EINVAL;
+	}
+
+	bq32002_lock_sem(dev);
+
+	/* Read time registers that contain some additional config/status bits */
+	err = i2c_burst_read_dt(&config->i2c, BQ32002_SECONDS_REG, regs, 3);
+	if (err) {
+		goto out_unlock;
+	}
+
+	/* Update the registers */
+	regs[0] = (regs[0] & ~BQ32002_SECONDS_MASK) |
+		  (bin2bcd(timeptr->tm_sec) & BQ32002_SECONDS_MASK);
+	regs[1] = bin2bcd(timeptr->tm_min) &
+		  BQ32002_MINUTES_MASK; /* Do not preserve oscillator fail flag, clear it */
+	regs[2] =
+		(regs[2] & ~BQ32002_HOURS_MASK) | (bin2bcd(timeptr->tm_hour) & BQ32002_HOURS_MASK);
```

Comment:
it looks like it would be reasonably trivial to handle years after 2100 by means of setting (and later checking) century bit (BQ32002_CENT_MASK) as needed?

Reply:
That's true. I thought about implementing that feature, but saw that `rtc_utils_validate_rtc_time` restricts the year value to 1900-2099 range: https://github.com/zephyrproject-rtos/zephyr/blob/4455cd7351ba2990aa6cb44bc09803b154d4a3f5/drivers/rtc/rtc_utils.c#L38

Not knowing exactly why there's such restriction I decided not to make it different just for this one driver

Reply:
@Lefucjusz oh i see -- although I guess you could still make the driver work in the 1900-2000 range though, by using the century bit :)

Reply:
Oh, you're right, I didn't think about this range being a use case for this but it is indeed! I'll add this feature later today :)

Reply:
The rtc utils validation checks for sanity as well, noone is setting an alarm, or the time, to after 2099 intentionally unless explicitly for unit testing :)

Reply:
@kartben I've added the century functionality :)